Add secret providers reference doc (#2126)
sensu/sensu-docs: content/sensu-go/5.17/reference/secret-providers.md

required     | true
type         | Integer
default      | 2
example      | {{< highlight shell >}}"max_retries": 2{{< /highlight >}}

timeout      | 
-------------|------ 
description  | Provider connection execution duration timeout (hard stop). In seconds.
required     | false
type         | Integer
default      | 60s
example      | {{< highlight shell >}}"timeout": "20s"{{< /highlight >}}

rate_limiter | 
-------------|------ 
description  | Maximum rate and burst limits for the secrets API.
required     | false
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"rate_limiter": {
  "limit": 10.0,
  "burst": 100
}
{{< /highlight >}}

#### Client attributes

address      | 
-------------|------ 
description  | Vault address to use for authentication.
required     | true
type         | String
example      | {{< highlight shell >}}
"address": "https://vaultserver.example.com:8200"
{{< /highlight >}}

token        | 
-------------|------ 
description  | Vault token to use for authentication.
required     | true
type         | String
example      | {{< highlight shell >}}"token": "VAULT_TOKEN"{{< /highlight >}}

version      | 
-------------|------ 

> description  | HashiCorp Vault version.

Author: ```suggestion
description  | HashiCorp Vault [version][14].
```

      },
      "max_retries": 2,
      "timeout": "20s",
      "rate_limiter": {
        "limit": 10.0,
        "burst": 100
      }
    }
  }
}
{{< /highlight >}}

{{< /language-toggle >}}

### Env example

Sensu's built-in `Env` secrets provider exposes secrets from backend [environment variables][4].
Using the `Env` secrets provider may require you to synchronize environment variables in Sensu backend clusters.

{{< language-toggle >}}

{{< highlight yml >}}
---
type: Env
api_version: secrets/v1
metadata:
  name: env
spec: {}
{{< /highlight >}}

{{< highlight json >}}
{
  "type": "Env",
  "api_version": "secrets/v1",
  "metadata": {
    "name": "env"
  },
  "spec": {}
}
{{< /highlight >}}

{{< /language-toggle >}}

[1]: ../../getting-started/enterprise/
[2]: ../../api/secrets/
[3]: ../../sensuctl/reference/
[4]: ../backend/#configuration-via-environment-variables
[5]: https://www.vaultproject.io/docs/what-is-vault/
[6]: ../../reference/rbac#default-users
[7]: ../../sensuctl/reference#create-resources
[8]: #spec-attributes
[9]: ../secrets/
[10]: https://www.vaultproject.io/docs/auth/token/
[11]: https://www.vaultproject.io/api/auth/cert/index.html
[12]: #client-attributes

> [13]: #env-example

Author: ```suggestion
[13]: #env-example
[14]: https://www.vaultproject.io/docs/commands/version/
```

    "version": "v1",
    "tls": {
      "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
    }
  },
  "max_retries": 2,
  "timeout": "20s",
  "rate_limiter": {
    "limit": 10.0,
    "burst": 100
  }
}
{{< /highlight >}}

### Metadata attributes

name         |      |
-------------|------
description  | Provider name used internally by Sensu.
required     | true
type         | String
example      | {{< highlight shell >}}"name": "vault"{{< /highlight >}}

### Spec attributes

client       | 
-------------|------ 
description  | Map that includes secrets provider configuration [client attributes][12].
required     | true
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"client": {
  "address": "https://vaultserver.example.com:8200",
  "token": "VAULT_TOKEN",
  "version": "v1",
  "tls": {
    "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
}
{{< /highlight >}}

max_retries  | 
-------------|------ 
description  | Number of times to retry connecting to the vault provider.
required     | true
type         | Integer
default      | 2
example      | {{< highlight shell >}}"max_retries": 2{{< /highlight >}}

timeout      | 
-------------|------ 

> description  | Provider connection execution duration timeout (hard stop). In seconds.

Reviewer: Let's simplify the description to `Provider connection timeout`

- The word execution obscures the meaning
- The units don't have to be seconds

For example, you could have "5h" or "100ms"

> description  | Provider connection execution duration timeout (hard stop). In seconds.

Author: ```suggestion
description  | Provider connection timeout (hard stop).
```

required     | false

> type         | Integer

Reviewer: Type here is String

> type         | Integer

Author: ```suggestion
type         | String
```

default      | 60s
example      | {{< highlight shell >}}"timeout": "20s"{{< /highlight >}}

rate_limiter | 
-------------|------ 
description  | Maximum rate and burst limits for the secrets API.
required     | false
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"rate_limiter": {
  "limit": 10.0,
  "burst": 100
}
{{< /highlight >}}

#### Client attributes

address      | 
-------------|------ 

> description  | Vault address to use for authentication.

Reviewer: Let's change the description to `Vault server address`.

The server is used for more than authentication.

> description  | Vault address to use for authentication.

Author: ```suggestion
description  | Vault server address.
```

required     | true
type         | String
example      | {{< highlight shell >}}
"address": "https://vaultserver.example.com:8200"
{{< /highlight >}}

token        | 
-------------|------ 
description  | Vault token to use for authentication.
required     | true
type         | String
example      | {{< highlight shell >}}"token": "VAULT_TOKEN"{{< /highlight >}}

version      | 
-------------|------ 
description  | HashiCorp Vault [version][14].
required     | true
type         | String
example      | {{< highlight shell >}}"version": "v1"{{< /highlight >}}

tls          | 
-------------|------ 

> description  | TLS object. Vault only works with TLS configured. You may need to set up a CA cert if it is not already stored in your operating system's trust store. To do this, set the TLS object, and provide the `ca_cert` path.

Author: ```suggestion
description  | TLS object. Vault only works with TLS configured. You may need to set up a CA cert if it is not already stored in your operating system's trust store. To do this, set the TLS object, and provide the `ca_cert` path. You may also need to set up a client cert, client key, or cname.
```

required     | true
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"tls": {

>   "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"

Reviewer: Some more example fields to add here are:

```
"client_cert": "/etc/ssl/certs/vault_cert.pem",
"client_key": "/etc/ssl/certs/vault_key.pem",
"cname": "vault_client.example.com"
```

>   "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"

Author: ```suggestion
  "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem",
  "client_cert": "/etc/ssl/certs/vault_cert.pem",
  "client_key": "/etc/ssl/certs/vault_key.pem",
  "cname": "vault_client.example.com"
```

}
{{< /highlight >}}

#### Rate limiter attributes

limit        | 
-------------|------ 
description  | Maximum number of secrets requests per second that can be transmitted to the backend with the agent secrets API.
required     | false

> type         | integer

Reviewer: Type is float here

> type         | integer

Author: ```suggestion
type         | Float
```

---
title: "Secrets providers"
linkTitle: "Secrets Providers"
description: "Sensu's secrets management capability allows you to avoid exposing secrets in your Sensu configuration. Read the reference to obtain secrets from one or more external secrets management providers and support references to external secrets in your Sensu configuration."
weight: 148
version: "5.17"
product: "Sensu Go"
menu: 
  sensu-go-5.17:
    parent: reference
---

- [Secrets providers specification](#secrets-providers-specification)
  - [Top-level attributes](#top-level-attributes) | [Metadata attributes](#metadata-attributes) | [Spec attributes](#spec-attributes)
- [Secrets providers examples](#secrets-providers-examples)

> **COMMERCIAL FEATURE**: Access the Env and Vault datatypes for secrets providers in the packaged Sensu Go distribution.

Reviewer 2: ```suggestion
**COMMERCIAL FEATURE**: Access the Env and VaultProvider datatypes for secrets providers in the packaged Sensu Go distribution.
```

> **COMMERCIAL FEATURE**: Access the Env and VaultProvider datatypes for secrets providers in the packaged Sensu Go distribution.

Reviewer 3: ```suggestion
**COMMERCIAL FEATURE**: Access the Env and VaultProvider secrets provider datatypes in the packaged Sensu Go distribution.
```

required     | true
type         | Integer
default      | 2
example      | {{< highlight shell >}}"max_retries": 2{{< /highlight >}}

timeout      | 
-------------|------ 
description  | Provider connection timeout (hard stop).
required     | false
type         | String
default      | 60s
example      | {{< highlight shell >}}"timeout": "20s"{{< /highlight >}}

rate_limiter | 
-------------|------ 
description  | Maximum rate and burst limits for the secrets API.
required     | false
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"rate_limiter": {
  "limit": 10.0,
  "burst": 100
}
{{< /highlight >}}

#### Client attributes

address      | 
-------------|------ 
description  | Vault server address.
required     | true
type         | String
example      | {{< highlight shell >}}
"address": "https://vaultserver.example.com:8200"
{{< /highlight >}}

token        | 
-------------|------ 
description  | Vault token to use for authentication.
required     | true
type         | String
example      | {{< highlight shell >}}"token": "VAULT_TOKEN"{{< /highlight >}}

version      | 
-------------|------ 

> description  | HashiCorp Vault [version][14].

Reviewer 2: The version refers to the API version, not the Vault version. The version can only be `v1` for production servers (see https://www.vaultproject.io/api-docs/#http-api) or `v2` for dev servers.

> description  | HashiCorp Vault [version][14].

Author: ```suggestion
description  | HashiCorp Vault [HTTP API version][14].
```

required     | true
type         | String
example      | {{< highlight shell >}}"version": "v1"{{< /highlight >}}

> tls          | 

Author: ```suggestion
<a name="tls-vault"></a>

tls          | 
```

-------------|------ 
description  | TLS object. Vault only works with TLS configured. You may need to set up a CA cert if it is not already stored in your operating system's trust store. To do this, set the TLS object, and provide the `ca_cert` path. You may also need to set up a client cert, client key, or cname.
required     | true
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"tls": {
  "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem",
  "client_cert": "/etc/ssl/certs/vault_cert.pem",
  "client_key": "/etc/ssl/certs/vault_key.pem",

>   "cname": "vault_client.example.com"

Reviewer 3: Unlike the other parameters in this TLS object, it's not immediately clear to me what the `cname` parameter does.

Author: Yes -- we do not seem to mention it elsewhere in our docs. @echlebek is it OK to remove the cname item?

Author: Update: I searched the Vault docs this morning and didn't find anything about 'cname' or 'canonical name'
I did find this: https://github.com/hashicorp/vault/issues/3709, so I guess it is an option for cert with Vault even though they don't describe it in their docs. Maybe we should keep it and link to https://en.wikipedia.org/wiki/CNAME_record?

Author: Update 2: then again, I don't see it in Cyril's PR: https://github.com/sensu/sensu-staging/pull/188/files/6c443c8f4afa36d898856453cf91d2c20d4dd47e..26f4ea6f851c4a66693d72af142619bdd8b7d610

So now I'm back to wondering if we can remove the cname item.

Reviewer: The cname corresponds to the name parameter mentioned here: https://www.vaultproject.io/api/auth/cert/index.html#parameters-7

Author: Got it, thank you. Do we need to explain why it is `cname` in our docs and `name` in the Vault docs?

}
{{< /highlight >}}

#### Rate limiter attributes

limit        | 
-------------|------ 

> description  | Maximum number of secrets requests per second that can be transmitted to the backend with the agent secrets API.

Reviewer 3: ```suggestion
description  | Maximum number of secrets requests per second that can be transmitted to the backend with the agent secrets API.
```

At the outset I assumed rate limiting would apply to communication with the Vault server, but "agent secrets API" makes it sound like we're configuring limits that apply to sensu-agents. Can we make this more clear?

Author: I believe it should just be "secrets API" (as in Sensu's secrets API).

{{< language-toggle >}}

{{< highlight yml >}}
---
type: VaultProvider
api_version: secrets/v1
metadata:
  name: vault1
spec:
  client:
    address: https://vaultserver.example.com:8200
    token: VAULT_TOKEN
    version: v1
    tls:
      ca_cert: "/etc/ssl/certs/vault_ca_cert.pem"
    max_retries: 2
    timeout: 20s
    rate_limiter:
      limit: 10
      burst: 100
{{< /highlight >}}

{{< highlight json >}}
{
  "type": "VaultProvider",
  "api_version": "secrets/v1",
  "metadata": {
    "name": "vault1"
  },
  "spec": {
    "client": {
      "address": "https://vaultserver.example.com:8200",
      "token": "VAULT_TOKEN",
      "version": "v1",
      "tls": {
        "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
      },
      "max_retries": 2,
      "timeout": "20s",
      "rate_limiter": {
        "limit": 10.0,
        "burst": 100
      }
    }
  }
}
{{< /highlight >}}

{{< /language-toggle >}}

### Env example

Sensu's built-in `Env` secrets provider exposes secrets from backend [environment variables][4].
Using the `Env` secrets provider may require you to synchronize environment variables in Sensu backend clusters.

> {{< language-toggle >}}

Reviewer 3: This very resource definition is very minimal, which makes sense for the Env provider, but it could lead to some confusion. Consider add a note explaining that the Env provider isn't configurable outside of metadata.name, but does need to be explicitly enabled by providing a config like this one (if that is in fact the case -- I'm just guessing).

Author: I am not sure either. I will follow up with Eric about this. Eric and I talked about creating a docs issue to add a guide for using Env-based secrets management too.

Author: I added Env to the secrets management guide in https://github.com/sensu/sensu-docs/pull/2130

@cwjohnston what do you think of referring to the guide for an example of how the Env provider is configured?

{{< highlight yml >}}
---
type: Env
api_version: secrets/v1
metadata:
  name: env
spec: {}
{{< /highlight >}}

{{< highlight json >}}
{
  "type": "Env",
  "api_version": "secrets/v1",
  "metadata": {
    "name": "env"
  },
  "spec": {}
}
{{< /highlight >}}

{{< /language-toggle >}}

[1]: ../../getting-started/enterprise/
[2]: ../../api/secrets/
[3]: ../../sensuctl/reference/
[4]: ../backend/#configuration-via-environment-variables
[5]: https://www.vaultproject.io/docs/what-is-vault/
[6]: ../../reference/rbac#default-users
[7]: ../../sensuctl/reference#create-resources
[8]: #spec-attributes
[9]: ../secrets/
[10]: https://www.vaultproject.io/docs/auth/token/
[11]: https://www.vaultproject.io/api/auth/cert/index.html
[12]: #client-attributes
[13]: #env-example

> [14]: https://www.vaultproject.io/docs/commands/version/

Author: ```suggestion
[14]: https://www.vaultproject.io/api-docs/
```

{{< /highlight >}}

token        | 
-------------|------ 
description  | Vault token to use for authentication.
required     | true
type         | String
example      | {{< highlight shell >}}"token": "VAULT_TOKEN"{{< /highlight >}}

version      | 
-------------|------ 
description  | HashiCorp Vault [HTTP API version][14].
required     | true
type         | String
example      | {{< highlight shell >}}"version": "v1"{{< /highlight >}}

<a name="tls-vault"></a>

tls          | 
-------------|------ 
description  | TLS object. Vault only works with TLS configured. You may need to set up a CA cert if it is not already stored in your operating system's trust store. To do this, set the TLS object, and provide the `ca_cert` path. You may also need to set up a client cert, client key, or cname.
required     | true
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"tls": {
  "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem",
  "client_cert": "/etc/ssl/certs/vault_cert.pem",
  "client_key": "/etc/ssl/certs/vault_key.pem",
  "cname": "vault_client.example.com"
}
{{< /highlight >}}

#### Rate limiter attributes

limit        | 
-------------|------ 

> description  | Maximum number of secrets requests per second that can be transmitted to the backend with the agent secrets API.

Author: ```suggestion
description  | Maximum number of secrets requests per second that can be transmitted to the backend with the secrets API.
```

linkTitle: "Secrets Providers"
description: "Sensu's secrets management capability allows you to avoid exposing secrets in your Sensu configuration. Read the reference to obtain secrets from one or more external secrets management providers and support references to external secrets in your Sensu configuration."
weight: 148
version: "5.17"
product: "Sensu Go"
menu: 
  sensu-go-5.17:
    parent: reference
---

- [Secrets providers specification](#secrets-providers-specification)
  - [Top-level attributes](#top-level-attributes) | [Metadata attributes](#metadata-attributes) | [Spec attributes](#spec-attributes)
- [Secrets providers examples](#secrets-providers-examples)

**COMMERCIAL FEATURE**: Access the Env and Vault datatypes for secrets providers in the packaged Sensu Go distribution.
For more information, see [Get started with commercial features][1].

Sensu's secrets management eliminates the need to expose secrets (e.g. usernames and passwords) in your Sensu configuration.
With Sensu's secrets management, you can obtain secrets from one or more external secrets providers, refer to external secrets, and consume secrets via backend [environment variables][4].

> Only Sensu backends have access to request [secrets][9] from a secrets provider.

Reviewer 2: Should this link be to the secrets reference rather than the secrets providers reference?

Author: Yes -- good catch, thank you!

required     | true
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"metadata": {
  "name": "vault"
}
{{< /highlight >}}

spec         | 
-------------|------
description  | Top-level map that includes secrets provider configuration [spec attributes][8].
required     | Required for secrets configuration in `wrapped-json` or `yaml` format.
type         | Map of key-value pairs
example      | {{< highlight shell >}}
"spec": {
  "client": {
    "address": "https://vaultserver.example.com:8200",
    "token": "VAULT_TOKEN",
    "version": "v1",
    "tls": {
      "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
    }
  },
  "max_retries": 2,
  "timeout": "20s",
  "rate_limiter": {
    "limit": 10.0,
    "burst": 100
  }
}
{{< /highlight >}}

### Metadata attributes

name         |      |
-------------|------
description  | Provider name used internally by Sensu.
required     | true
type         | String
example      | {{< highlight shell >}}"name": "vault"{{< /highlight >}}

### Spec attributes

client       | 
-------------|------ 
description  | Map that includes secrets provider configuration [client attributes][12].

> required     | NEEDED

Reviewer 2: ```suggestion
required     | true
```

type         | Map of key-value pairs
example      | {{< highlight shell >}}
"client": {
  "address": "https://vaultserver.example.com:8200",
  "token": "VAULT_TOKEN",
  "version": "v1",
  "tls": {
    "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
}
{{< /highlight >}}

max_retries  | 
-------------|------ 
description  | Number of times to retry connecting to the vault provider.
required     | true
type         | Integer

> default      | NEEDED

Reviewer 2: ```suggestion
default      | 2
```

example      | {{< highlight shell >}}"max_retries": 2{{< /highlight >}}

timeout      | 
-------------|------ 
description  | Provider connection execution duration timeout (hard stop). In seconds.

> required     | true

Reviewer 2: ```suggestion
required     | false
```

type         | Integer

> default      | `false`

Reviewer 2: Looks like the default timeout is 60s.

> default      | `false`

Author: ```suggestion
default      | 60s
```

example      | {{< highlight shell >}}"timeout": "20s"{{< /highlight >}}

rate_limiter | 
-------------|------ 
description  | Maximum rate and burst limits for the secrets API.

> required     | true

Reviewer 2: ```suggestion
required     | false
```

type         | Map of key-value pairs

> default      | NEEDED

Reviewer 2: No default.

> default      | NEEDED

Author: ```suggestion
```

example      | {{< highlight shell >}}
"rate_limiter": {
  "limit": 10.0,
  "burst": 100
}
{{< /highlight >}}

#### Client attributes

address      | 
-------------|------ 
description  | Vault address to use for authentication.
required     | true
type         | String

> default      | NEEDED

Reviewer 2: No default.

Reviewer 2: Not for this field!

> default      | NEEDED

Author: ```suggestion
```

example      | {{< highlight shell >}}
"address": "https://vaultserver.example.com:8200"
{{< /highlight >}}

token        | 
-------------|------ 
description  | Vault token to use for authentication.
required     | true
type         | String

> default      | NEEDED

Reviewer 2: No default.

Author: Should I change "required" to "false"?

> default      | NEEDED

Author: ```suggestion
```

example      | {{< highlight shell >}}"token": "VAULT_TOKEN"{{< /highlight >}}

version      | 
-------------|------ 

> description  | HashiCorp Vault version.

Reviewer 2: We should include a link to the HashiCorp Vault docs that specify where to find this.. the value can only be `v1` for production or `v2` for dev.

required     | true
type         | String

> default      | NEEDED

Reviewer 2: No default.

Author: Should "required" be changed to "false"?

> default      | NEEDED

Author: ```suggestion
```

example      | {{< highlight shell >}}"version": "v1"{{< /highlight >}}

tls          | 
-------------|------ 

> description  | TLS object. Vault only works with TLS configured. You may need to set up a CA cert if it is not already stored in your operating system's trust store. To do this, set the TLS object, and provide the `ca_cert` path.

Reviewer 2: @echlebek is this correct? I thought dev vault servers don't require TLS.

Reviewer: Any real use of Vault is going to require TLS, I don't think dev servers count.

required     | true
type         | Map of key-value pairs

> default      | NEEDED

Reviewer 2: No default.

Author: Change "required" to "false"?

> default      | NEEDED

Author: ```suggestion
```

example      | {{< highlight shell >}}
"tls": {
  "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
}
{{< /highlight >}}

#### Rate limiter attributes

limit        | 
-------------|------ 
description  | Maximum number of secrets requests per second that can be transmitted to the backend with the agent secrets API.

> required     | NEEDED

Reviewer 2: ```suggestion
required     | false
```

type         | integer

> default      | NEEDED

Reviewer 2: No default for optional fields.

> default      | NEEDED

Author: ```suggestion
```

example      | {{< highlight shell >}}"limit": 10.0{{< /highlight >}}

burst        | 
-------------|------ 
description  | Maximum amount of burst allowed in a rate interval for the secrets API.

> required     | NEEDED

Reviewer 2: ```suggestion
required     | false
```

type         | integer

> default      | NEEDED

Reviewer 2: No default for optional fields.

> default      | NEEDED

Author: ```suggestion
```

      "token": "VAULT_TOKEN",
      "version": "v1",
      "tls": {
        "ca_cert": "/etc/ssl/certs/vault_ca_cert.pem"
      },
      "max_retries": 2,
      "timeout": "20s",
      "rate_limiter": {
        "limit": 10.0,
        "burst": 100
      }
    }
  }
}
{{< /highlight >}}

{{< /language-toggle >}}

### Env example

Sensu's built-in `Env` secrets provider exposes secrets from backend [environment variables][4].
Using the `Env` secrets provider may require you to synchronize environment variables in Sensu backend clusters.

{{< language-toggle >}}

{{< highlight yml >}}
---
type: Env
api_version: secrets/v1
metadata:
  name: env
spec: {}
{{< /highlight >}}

{{< highlight json >}}
{
  "type": "Env",
  "api_version": "secrets/v1",
  "metadata": {
    "name": "env"
  },
  "spec": {}
}
{{< /highlight >}}

{{< /language-toggle >}}

[1]: ../../getting-started/enterprise/
[2]: ../../api/secrets/
[3]: ../../sensuctl/reference/
[4]: ../backend/#configuration-via-environment-variables
[5]: https://www.vaultproject.io/docs/what-is-vault/
[6]: ../../reference/rbac#default-users
[7]: ../../sensuctl/reference#create-resources
[8]: #spec-attributes

> [9]: ../secret-providers/

Author: ```suggestion
[9]: ../secrets/
```
